python_modules/libraries/dagster-aws/dagster_aws/s3/io_manager.py

```diff
@@ -1,107 +1,80 @@
-        check.param_invariant(len(key) > 0, "key")
-
-        # delete_object wont fail even if the item has been deleted.
-        self.s3.delete_object(Bucket=self.bucket, Key=key)
-
-    def _has_object(self, key):
-        check.str_param(key, "key")
-        check.param_invariant(len(key) > 0, "key")
-
-        found_object = False
+        pickled_obj = pickle.dumps(obj, PICKLE_PROTOCOL)
+        pickled_obj_bytes = io.BytesIO(pickled_obj)
+        self.s3.upload_fileobj(pickled_obj_bytes, self.bucket, str(path))
 
+    def path_exists(self, path: UPath) -> bool:
         try:
-            self.s3.get_object(Bucket=self.bucket, Key=key)
-            found_object = True
+            self.s3.get_object(Bucket=self.bucket, Key=str(path))
         except self.s3.exceptions.NoSuchKey:
-            found_object = False
+            return False
+        return True
 
-        return found_object
+    def get_loading_input_log_message(self, path: UPath) -> str:
+        return f"Loading S3 object from: {self._uri_for_path(path)}"
 
-    def _uri_for_key(self, key):
-        check.str_param(key, "key")
-        return "s3://" + self.bucket + "/" + f"{key}"
+    def get_writing_output_log_message(self, path: UPath) -> str:
+        return f"Writing S3 object at: {self._uri_for_path(path)}"
 
-    def load_input(self, context):
-        if context.dagster_type.typing_type == type(None):
-            return None
+    def path_rm(self, path: UPath) -> None:
+        self.s3.delete_object(Bucket=self.bucket, Key=str(path))
 
-        key = self._get_path(context)
-        context.log.debug(f"Loading S3 object from: {self._uri_for_key(key)}")
-        obj = pickle.loads(self.s3.get_object(Bucket=self.bucket, Key=key)["Body"].read())
+    def path_mkdir(self, path: UPath) -> None:
+        # It is not necessary to create directories in S3
+        return None
 
-        return obj
+    def get_metadata(self, context: OutputContext, obj: Any) -> Dict[str, MetadataValue]:
+        path = self._get_path(context)
+        return {"uri": MetadataValue.path(self._uri_for_path(path))}
 
-    def handle_output(self, context, obj):
-        if context.dagster_type.typing_type == type(None):
-            check.invariant(
-                obj is None,
-                (
-                    "Output had Nothing type or 'None' annotation, but handle_output received"
-                    f" value that was not None and was of type {type(obj)}."
-                ),
-            )
-            return None
-
-        key = self._get_path(context)
-        path = self._uri_for_key(key)
-        context.log.debug(f"Writing S3 object at: {path}")
-
-        if self._has_object(key):
-            context.log.warning(f"Removing existing S3 key: {key}")
-            self._rm_object(key)
+    def get_op_output_relative_path(self, context: Union[InputContext, OutputContext]) -> UPath:
+        return UPath("storage", super().get_op_output_relative_path(context))
```

Comment:
Why do we need to add "storage" here, but not in the same method for the fs_io_manager?

Reply:
I'm just preserving existing behavior here-- the status quo S3 IO manager uses "storage", but FS IO manager doesn't. Don't know why.

Reply:
What's puzzling me here is the docstring for `fs_io_manager`, where we mention a "storage" prefix. But it looks like that gets added in `LocalArtifactStorage`. So this looks fine.